### PR TITLE
Add high virtuality limit to Python interface

### PR DESCRIPTION
**Pull request by felixhekhorn**
- the `bmsn` module, which contains the actual limit (and was added already before this PR), is copied from [`yadism`](https://github.com/NNPDF/yadism/blob/9617329a4025a27ac6de99c0ba904e276fcb8dfb/src/yadism/coefficient_functions/fonll/raw_nc.py)
- the strategy is the same as for the threshold limit ($\eta\to 0$):
  - if $\xi > \xi_{max}$: use only limit
  - elif $\xi_{max} > \xi > \xi_{mix}$: interpolate limit and exact computation
  - else: use exact computation
- as now two limits are in game, I might have generated a discontinuity for $\eta\to 0,\xi\to\infty$ - as I'm lazy, I keep it until it's proven it is a problem
- with this version we can do `LeProHQ.cg1("F2","VV",10000000,1)` which returns a number instead of an exception

@niclaurenti can you please have a look?

@giacomomagni here I added only $F_2$ and $F_L$ and I have to repeat this for $g_1$ - do we have the limit already coded somewhere?

cc @AleCandido

### CODE REVIEW

**Comment by alecandido** [comment]
At some point I'd consider dropping `packutil`, since we're not maintaining it...

True that there is not even anything broken, nor it's going to break in the close future. But we'd never know until it will.

File: Benchmarks/NPB472-611/test.py
Comment (alecandido) [inline Benchmarks/NPB472-611/test.py:1]: You could even consider uploading the plot on *the* website :P

**Comment by felixhekhorn** [comment]
> At some point I'd consider dropping `packutil`, since we're not maintaining it...
> 
> True that there is not even anything broken, nor it's going to break in the close future. But we'd never know until it will.

yeah, I know - I should move to poetry ... but not sure how much mess it will create :see_no_evil:

File: LeProHQpy/src/LeProHQ/cq1_.py [hunk 1]
```diff
@@ -36,7 +36,17 @@ def cq1t(proj, cc, xi, eta):
         * (a11 * np.log(beta) + a10)
     )
 
+def cq1hv(proj,cc,xi,eta):
+    """High virtuality limit of cq1."""
+    l = np.log(xi)
+    z = xi / (4.*(1.+eta) + xi)
+    n = xi * 4**2 * np.pi / z
+    if "F2" == proj:
```
Comment (alecandido) [inline LeProHQpy/src/LeProHQ/cq1_.py]: It is customary to write it in the inverse fashion:
```suggestion
    if proj == "F2":
```

File: LeProHQpy/src/LeProHQ/cq1_.py [hunk 2]
```diff
@@ -36,7 +36,17 @@ def cq1t(proj, cc, xi, eta):
...
+        return (bmsn.c2ps2am0_aq2(z) * l**2 + bmsn.c2ps2am0_aq(z) * l + bmsn.c2ps2am0_a0(z)) / n
+    elif "FL" == proj:
```
Comment (alecandido) [inline LeProHQpy/src/LeProHQ/cq1_.py]: See above

File: LeProHQpy/src/LeProHQ/cq1_.py [hunk 3]
```diff
@@ -2,11 +2,11 @@
 import numba as nb
 import numpy as np
 
-from .cg0 import cg0t
+from .cg0_ import cg0t
 from .color import Kgph, Kqph
 from .partonic_vars import build_eta, build_xi
 from .utils import ln2, raw_c
-
+from . import bmsn
```
Comment (alecandido) [inline LeProHQpy/src/LeProHQ/cq1_.py:9]: I actually don't see this file...

Comment (alecandido) [reply LeProHQpy/src/LeProHQ/cq1_.py:9]: Nor a variable with this name defined in `__init__.py`

File: Benchmarks/NPB472-611/test.py
Comment (felixhekhorn) [reply Benchmarks/NPB472-611/test.py:1]: well, at some point I had the idea of expose all my benchmarks on a github-pages (to show off :innocent: ) - but of course this never happened due to low priority

**Comment by alecandido** [comment]
> yeah, I know - I should move to poetry ... but not sure how much mess it will create see_no_evil

You don't have necessarily to embrace Poetry. By now, people are doing something similar to what we were doing with `packutil` just using [`importlib.metadata`](https://docs.python.org/3/library/importlib.metadata.html) (new in py3.8, but we dropped support for py3.7...).

File: LeProHQpy/src/LeProHQ/cq1_.py (hunk 3 above)
Comment (felixhekhorn) [reply LeProHQpy/src/LeProHQ/cq1_.py:9]: as said, is already on [master](https://github.com/felixhekhorn/LeProHQ/blob/d3566cfd675fdcdfa4a3a1223d045ffd69b32dc5/LeProHQpy/src/LeProHQ/bmsn.py)

Comment (alecandido) [reply LeProHQpy/src/LeProHQ/cq1_.py:9]: Ah, you're right, I quickly passed through that sentence and misunderstood the meaning, sorry

File: LeProHQpy/src/LeProHQ/cq1_.py (hunk 1 above)
Comment (felixhekhorn) [reply LeProHQpy/src/LeProHQ/cq1_.py]: I did fall back to C style - reverted in [ab52f96](https://github.com/felixhekhorn/LeProHQ/pull/2/commits/ab52f967cffa0af30fb3793a33a12c12a47e6540)

**Comment by alecandido** [comment]
In principle all of this could have been done directly in `yadism`, saving some duplication (we'd just needed to catch the exception).

But `yadism` is currently a mess, and it's going to be reworked. So, it's fine.
However, since the interpolation is actually done in a range, I wonder if you should add a raw mode to still raise the exception. But maybe it's just useless...

**Comment by felixhekhorn** [comment]
> In principle all of this could have been done directly in `yadism`, saving some duplication

true, but then I would argue the correct place is here and `yadism` can reuse as I'm more general here, moreover `yadism` would need to act in two places (skipping both heavy and asy) - in any case, I agree: let's not worry now ...

> I wonder if you should add a raw mode to still raise the exception

yes, I was thinking the same ... but that would be not public in any case ... (simlar as the limits can be accessed on their own)

**Comment by niclaurenti** [comment]
> * as now two limits are in game, I might have generated a discontinuity for η→0,ξ→∞ - as I'm lazy, I keep it until it's proven it is a problem

What happens when $\eta \rightarrow 0$ and $\xi \rightarrow \infty$? I mean, which limit is returned?

**Comment by felixhekhorn** [comment]
> What happens when η→0 and ξ→∞? I mean, which limit is returned?

1. if $\eta < \eta_{min}$: return the threshold limit
2. elif $\xi > \xi_{max}$: return the high virtuality limit
3. else ...

so when $\eta = \eta_{min}$ you switch between the two limits and they are not smooth

**Comment by alecandido** [comment]
For smoothness you'd need a combined expression (I've already seen this...).

However, I believe that the discontinuity should happen in a not very relevant region, and most likely is not uniformly bad.
In the 2D (x, Q2) plane, the discontinuity will happen on a line, and it will get worse and worse the more you go far from the interpolated region.

Can you identify the "bad region" in that plane?

**Comment by niclaurenti** [comment]
Anyway I don't think that this non smoothness will give a high impact: first of all, I've seen that in `FiatLux` cutting the integral at `q2=60000` gives an error of the order of percent, so this jump will be probably a "percent of the percent". Second, I don't know exactly the values of `eta_min` but I guess it is in the very small x region (for the fortran module that I used in my thesis it was for $x \simeq 10^{-6}$ and for this code I think it is in a similar region), where things are not reliable in any case.
However, if this branch is working I can use it  to produce a grid and see if the plots are good or not.

**Comment by felixhekhorn** [comment]
> Can you identify the "bad region" in that plane?

the discontinuity is at $\eta = \eta_{min},\xi>\xi_{max}$

> I don't know exactly the values of eta_min but I guess it is in the very small x region

$\eta_{min} = 0.1$ for the gluon and below that point I return the fully analytic exact result - I believe the old Fortran implementation it is a full interpolation which will just stop at some point

> probably a "percent of the percent"

I expect "a permille of the percent" or less - it is a single line in a small phase space corner ...

**Pull request merged**